### ISSUE TITLE
[S028] Connect model logits to targets and prove causal behavior

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2#1]
**Prerequisites:** S010, S026–S027; #1 applies. **Timebox:** 35–50 minutes.

## Learn and implement
Add a small loss-evaluation entry point around the existing forward and loss functions in `src/gpt2.c`; create `tests/test_model_loss.c`. Keep targets out of the model's forward computation: they are used only after logits exist. Validate the whole request first. Document that logits at t predict the already-shifted target at t; do not shift a second time inside loss.

## Explicit tests
Use S010's known batch alignment and all-zero parameters: mean loss must be `log(V)` for every legal target set. Change targets without changing inputs: logits must remain identical. With fixed nonzero parameters, change an input token at position j; logits at all `t<j` must stay within #1 tolerances. Change only the second batch row: first-row logits must be unchanged. Repeating a batch must preserve mean loss. Reject bad targets, `T>P`, and mismatched workspace shapes before unsafe access.
```sh
make test TEST=model_loss
make sanitize TEST=model_loss
make test
```
All pass. Temporarily remove the causal restriction in attention and confirm the future-token test fails on a nondegenerate fixture. A zero-weight fixture alone cannot test causality.

## Done / teach-back
Explain why using future targets to build hidden states would invalidate next-token training. Show the exact `(input position,target position)` pairing. No optimizer or learning yet.
